interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp

```diff
@@ -433,29 +507,99 @@ namespace cling {
+
+          // Handle lib dependencies
+          std::string RPath;
+          std::string RunPath;
+          std::vector<StringRef> Deps;
+          llvm::StringRef NewLibP = NewLibPath->GetFullName();
+          auto ObjFileOrErr =
+              llvm::object::ObjectFile::createObjectFile(NewLibP);
+          if (llvm::Error Err = ObjFileOrErr.takeError()) {
+            if (DEBUG > 1) {
+              std::string Message;
+              handleAllErrors(std::move(Err), [&](llvm::ErrorInfoBase& EIB) {
+                Message += EIB.message() + "; ";
+              });
+              cling::errs()
+                  << "Dyld::ScanForLibraries: Failed to read object file "
+                  << NewLibP << " Errors: " << Message << "\n";
+            }
+            return nullptr;
+          }
+          llvm::object::ObjectFile* BinObjF = ObjFileOrErr.get().getBinary();
+          if (BinObjF->isELF()) {
+            if (const auto* ELF = dyn_cast<ELF32LEObjectFile>(BinObjF))
+              HandleDynTab(ELF->getELFFile(), FileName, RPath, RunPath, Deps);
+            else if (const auto* ELF = dyn_cast<ELF32BEObjectFile>(BinObjF))
+              HandleDynTab(ELF->getELFFile(), FileName, RPath, RunPath, Deps);
+            else if (const auto* ELF = dyn_cast<ELF64LEObjectFile>(BinObjF))
+              HandleDynTab(ELF->getELFFile(), FileName, RPath, RunPath, Deps);
+            else if (const auto* ELF = dyn_cast<ELF64BEObjectFile>(BinObjF))
+              HandleDynTab(ELF->getELFFile(), FileName, RPath, RunPath, Deps);
+
+          } else if (BinObjF->isMachO()) {
+            MachOObjectFile* Obj = (MachOObjectFile*)BinObjF;
+            for (const auto& Command : Obj->load_commands()) {
+              if (Command.C.cmd == MachO::LC_LOAD_DYLIB) {
+                // Command.C.cmd == MachO::LC_ID_DYLIB ||
+                // Command.C.cmd == MachO::LC_LOAD_WEAK_DYLIB ||
+                // Command.C.cmd == MachO::LC_REEXPORT_DYLIB ||
+                // Command.C.cmd == MachO::LC_LAZY_LOAD_DYLIB ||
+                // Command.C.cmd == MachO::LC_LOAD_UPWARD_DYLIB ||
+                MachO::dylib_command dylibCmd =
+                    Obj->getDylibIDLoadCommand(Command);
+                Deps.push_back(StringRef(Command.Ptr + dylibCmd.dylib.name));
+              } else if (Command.C.cmd == MachO::LC_RPATH) {
+                MachO::rpath_command rpathCmd = Obj->getRpathCommand(Command);
+                CombinePaths(RPath, Command.Ptr + rpathCmd.path);
+              }
+            }
+          }
+
+          for (StringRef dep : Deps) {
+            std::string dep_full =
+                m_DynamicLibraryManager.lookupLibrary(dep, RPath, RunPath);
+            const LibraryPath* DepLibPath = HandleLib(dep_full, level + 1);
+            // NewLibPath. add dependency lib to dep list?
+            if (!DepLibPath) {
+              const_cast<LibraryPath*>(NewLibPath)
+                  ->m_LibDeps.push_back(DepLibPath);
```

Comment:
For experiments, with ten thousand libraries, and path lengths of 1k, and in average 20 dependent libraries each: how much memory are we talking about here?

Reply:
See previous comment.

Reply:
Could you please check with Shahzad for a profile before / after this PR is merged, both in *realtime* (due to many file ops) and *memory*?

Reply:
@davidlange6, could you give us a hand?

Reply:
The stats from cmssw look good.